lib/host/wasi/inode-macos.cpp

```diff
@@ -1111,7 +1111,119 @@ WasiExpect<void> Poller::wait(CallbackType Callback) noexcept {
   }
   return {};
 }
+WasiExpect<void>
+INode::getAddrinfo(const char *NodeStr, const char *ServiceStr,
+                   const __wasi_addrinfo_t &Hint, uint32_t MaxResLength,
+                   std::vector<struct __wasi_addrinfo_t *> *WasiAddrinfoArray,
+                   std::vector<struct __wasi_sockaddr_t *> *WasiSockaddrArray,
+                   std::vector<char *> *AiAddrSaDataArray,
+                   std::vector<char *> *AiCanonnameArray,
+                   /*Out*/ __wasi_size_t *ResLength) noexcept {
+  struct addrinfo SysHint;
+  struct addrinfo *SysResPtr = nullptr;
+  struct addrinfo *SysResItem = nullptr;
+  SysHint.ai_flags = Hint.ai_flags;
+  SysHint.ai_family = Hint.ai_family;
+  SysHint.ai_socktype = Hint.ai_socktype;
+  SysHint.ai_protocol = Hint.ai_protocol;
+  SysHint.ai_addrlen = Hint.ai_addrlen;
+  SysHint.ai_addr = NULL;
+  SysHint.ai_canonname = NULL;
+  SysHint.ai_next = NULL;
+
+  int POSIXReturn = ::getaddrinfo(NodeStr, ServiceStr, &SysHint, &SysResPtr);
+  switch (POSIXReturn) {
+  case EAI_ADDRFAMILY:
+    return WasiUnexpect(__WASI_ERRNO_AIADDRFAMILY);
+  case EAI_AGAIN:
+    return WasiUnexpect(__WASI_ERRNO_AIAGAIN);
+  case EAI_BADFLAGS:
+    return WasiUnexpect(__WASI_ERRNO_AIBADFLAG);
+  case EAI_FAIL:
+    return WasiUnexpect(__WASI_ERRNO_AIFAIL);
+  case EAI_FAMILY:
+    return WasiUnexpect(__WASI_ERRNO_AIFAMILY);
+  case EAI_MEMORY:
+    return WasiUnexpect(__WASI_ERRNO_AIMEMORY);
+  case EAI_NODATA:
+    return WasiUnexpect(__WASI_ERRNO_AINODATA);
+  case EAI_NONAME:
+    return WasiUnexpect(__WASI_ERRNO_AINONAME);
+  case EAI_SERVICE:
+    return WasiUnexpect(__WASI_ERRNO_AISERVICE);
+  case EAI_SOCKTYPE:
+    return WasiUnexpect(__WASI_ERRNO_AISOCKTYPE);
+  case EAI_SYSTEM:
+    return WasiUnexpect(__WASI_ERRNO_AISYSTEM);
+  }
+  // calculate ResLength
+  auto calculateResSize = [](struct addrinfo *const Res) -> __wasi_size_t {
+    struct addrinfo *TmpPointer;
+    __wasi_size_t Length = 0;
+    for (TmpPointer = Res; TmpPointer != nullptr;
+         TmpPointer = TmpPointer->ai_next) {
+      Length++;
+    }
+    return Length;
+  };
+
+  *ResLength = calculateResSize(SysResPtr);
+  if (*ResLength > MaxResLength) {
+    *ResLength = MaxResLength;
+  }
+
+  SysResItem = SysResPtr;
+
+  for (uint32_t Idx = 0; Idx < (*ResLength); Idx++) {
+    (*WasiAddrinfoArray)[Idx]->ai_flags =
+        static_cast<__wasi_aiflags_t>(SysResItem->ai_flags);
+    (*WasiAddrinfoArray)[Idx]->ai_socktype =
+        static_cast<__wasi_sock_type_t>(SysResItem->ai_socktype);
+    (*WasiAddrinfoArray)[Idx]->ai_protocol =
+        static_cast<__wasi_protocol_t>(SysResItem->ai_protocol);
+    (*WasiAddrinfoArray)[Idx]->ai_family =
+        static_cast<__wasi_address_family_t>(SysResItem->ai_family);
+    (*WasiAddrinfoArray)[Idx]->ai_addrlen = SysResItem->ai_addrlen;
```

Comment:
Use reference to access this element, don't use index access on the same element every time.